libgraphics/src/drawing.cpp

```diff
@@ -0,0 +1,1894 @@
+        r_singleton = *t_singleton_ptr;
+        
+        m_bytes = t_singleton_ptr + 1;
+        m_byte_count -= t_singleton_size;
+        
+        return true;
+    }
+    
+    template<typename T>
+    bool Array(size_t p_count, T*& r_array)
+    {
+        size_t t_array_size = p_count * sizeof(T);
+        
+        if (m_byte_count < t_array_size)
+        {
+            return false;
+        }
+        
+        auto t_array_ptr = static_cast<const T *>(m_bytes);
+        r_array = t_array_ptr;
+        
+        m_bytes = t_array_ptr + p_count;
+        m_byte_count -= t_array_size;
+        
+        return true;
+    }
+
+private:
+    const void *m_bytes;
+    size_t m_byte_count;
+};
+
+class MCGDrawingContext
+{
+public:
+    MCGDrawingContext(const void *p_drawing_bytes, size_t p_drawing_byte_count)
+    {
+        MCGDrawingByteStream t_stream(p_drawing_bytes, p_drawing_byte_count);
+        
+        /* Unpack the header. If the size of the provided data does not match
+         * what is expected by unpacking the data, then it is an
+         * 'InvalidDrawing' */
+        uint8_t t_ident_0, t_ident_1, t_ident_2, t_ident_version;
+        uint32_t t_flags;
+        if (!t_stream.Singleton(t_ident_0) ||
+            !t_stream.Singleton(t_ident_1) ||
+            !t_stream.Singleton(t_ident_2) ||
+            !t_stream.Singleton(t_ident_version) ||
+            !t_stream.Singleton(t_flags) ||
+            !t_stream.Singleton(m_scalar_count) ||
+            !t_stream.Array(m_scalar_count, m_scalars) ||
+            !t_stream.Singleton(m_opcode_count) ||
+            !t_stream.Array(m_opcode_count, m_opcodes) ||
+            !t_stream.IsFinished())
+        {
+            m_status = kMCGDrawingStatusInvalidDrawing;
+            return;
+        }
+        
+        /* Check the ident field is correct. */
+        if (t_ident_0 != kMCGDrawingIdent0 ||
+            t_ident_1 != kMCGDrawingIdent1 ||
+            t_ident_2 != kMCGDrawingIdent2)
+        {
+            m_status = kMCGDrawingStatusInvalidIdent;
+            return;
+        }
+        
+        /* Check the version field is correct. */
+        if (t_ident_version != kMCGDrawingVersion)
+        {
+            m_status = kMCGDrawingStatusInvalidVersion;
+            return;
+        }
+        
+        /* Check the flags field is correct. */
+        if ((t_flags & ~kMCGDrawingHeaderFlagValidMask) != 0)
+        {
+            m_status = kMCGDrawingStatusInvalidFlags;
+            return;
+        }
+        
+        /* Process the width, if present. If the width is not present, then
+         * it is taken to be '100%' (-1). */
+        if ((t_flags & kMCGDrawingHeaderFlagHasWidthBit) != 0)
+        {
+            if (m_scalar_count < 1)
+            {
+                m_status = kMCGDrawingStatusInvalidWidth;
+                return;
+            }
+            m_width = m_scalars[0];
+            m_scalars += 1;
+            m_scalar_count -= 1;
```

Comment:
By no means necessary but maybe a little `GetScalar` helper function would be good here that updates the scalars params and returns the scalar?

Reply:
I'm going to change this so init uses the PC/SC as elsewhere - it will leave them in the state for rendering the main drawing.